hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,410 @@
+ *     else write the record as is
+ * For all pending records from incoming batch, write to file.
+ *
+ * Illustration with simple data.
+ * Incoming data:
+ *     rec1_2, rec4_2, rec5_1, rec6_1
+ * Existing data:
+ *     rec1_1, rec2_1, rec3_1, rec4_1
+ *
+ * For every existing record, merge w/ incoming if required and write to storage.
+ *    => rec1_1 and rec1_2 is merged to write rec1_2 to storage
+ *    => rec2_1 is written as is
+ *    => rec3_1 is written as is
+ *    => rec4_2 and rec4_1 is merged to write rec4_2 to storage
+ * Write all pending records from incoming set to storage
+ *    => rec5_1 and rec6_1
+ *
+ * Final snapshot in storage
+ * rec1_2, rec2_1, rec3_1, rec4_2, rec5_1, rec6_1
+ *
+ * </p>
+ */
+@NotThreadSafe
+public class HoodieSparkMergeHandleV2<T, I, K, O> extends HoodieWriteHandle<T, I, K, O> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieMergeHandle.class);
+
+  protected Set<String> writtenRecordKeys;
+  protected HoodieReaderContext readerContext;
+  protected HoodieFileWriter fileWriter;
+  protected boolean preserveMetadata = false;
+
+  protected StoragePath newFilePath;
+  protected StoragePath oldFilePath;
+  protected long recordsWritten = 0;
+  // TODO(yihua): audit delete stats because file group reader may not return deletes
+  protected long recordsDeleted = 0;
+  protected long updatedRecordsWritten = 0;
+  protected long insertRecordsWritten = 0;
+  protected Option<BaseKeyGenerator> keyGeneratorOpt;
+  protected FileSlice fileSlice;
+  private HoodieBaseFile baseFileToMerge;
+
+  protected Option<String[]> partitionFields = Option.empty();
+  protected Object[] partitionValues = new Object[0];
+
+  /**
+   * Called by compactor code path using the file group reader.
+   */
+  public HoodieSparkMergeHandleV2(HoodieWriteConfig config, String instantTime, HoodieTable<T, I, K, O> hoodieTable,
+                                  String partitionPath, String fileId,
+                                  CompactionOperation operation, TaskContextSupplier taskContextSupplier, Option<BaseKeyGenerator> keyGeneratorOpt,
+                                  HoodieReaderContext readerContext) {
+    super(config, instantTime, partitionPath, fileId, hoodieTable, taskContextSupplier);
+    this.readerContext = readerContext;
+    Option<HoodieBaseFile> baseFileOpt =
+        operation.getBaseFile(config.getBasePath(), operation.getPartitionPath());
+    List<HoodieLogFile> logFiles = operation.getDeltaFileNames().stream().map(p ->
+            new HoodieLogFile(new StoragePath(FSUtils.constructAbsolutePath(
+                config.getBasePath(), operation.getPartitionPath()), p)))
+        .collect(Collectors.toList());
+    this.fileSlice = new FileSlice(
+        operation.getFileGroupId(),
+        operation.getBaseInstantTime(),
+        baseFileOpt.isPresent() ? baseFileOpt.get() : null,
+        logFiles);
+    ValidationUtils.checkArgument(baseFileOpt.isPresent(),
```

Comment:
does the writer path prevent this from happening? i.e encoding partial updates without a base file. what does this even mean?

Reply:
While I was working on the PoC, this is a reminder for myself to implement compaction for log files only (no base file), which should go through `HoodieSparkMergeHandleV2`.  The old code path uses `HoodieCreateHandle` to handle log only merging (i.e., `HoodieCompactionHandler#handleInsert`).

I've added the compaction support based on the file group reader for log files only (no base file) and removed the check.